homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -65,53 +76,103 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
     envoy_reader = EnvoyReader(ip_address, username, password)
 
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        async with async_timeout.timeout(30):
+            i = 0
+            while i < 3:
+                try:
+                    data = await envoy_reader.update()
```

Comment:
Can this logic be moved into the upstream library?

Reply:
I don't think I even need to have it in the envoy_reader library. The sensor retrieves data every 60seconds. Envoy hardware devices update anywhere from real-time to up-to every 15mins, depending on the model. If one polling cycle is missed due to some network communication problem or maybe the device can't handle the request (which I hope won't happen as often anymore after this PR gets merged!).

I can remove the while loop and just raise UpdateFailed(...).

Reply:
@bdraco made my code change